pkg/assetfetcher/http.go

```diff
@@ -0,0 +1,164 @@
+package assetfetcher
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/asecurityteam/nexpose-vuln-notifier/pkg/domain"
+)
+
+// This const block helps format our request to the Nexpose sites asset endpoint:
+// GET /api/3/sites/{id}/assets (doc: https://help.rapid7.com/insightvm/en-us/api/index.html#operation/getSiteAssets)
+const (
+	basePath       = "/api/3"
+	nexposeSite    = "/sites"
+	nexposeAssets  = "/assets"
+	pageQueryParam = "page" // The index of the page (zero-based) to retrieve.
+	sizeQueryParam = "size" // The number of records per page to retrieve.
+)
+
+// SiteAssetsResponse is the structure of the Nexpose site assets response
+type SiteAssetsResponse struct {
+	// Hypermedia links to corresponding or related resources
+	Links domain.Link
+	// The details of pagination indicating which page was returned, and how the remaining pages can be retrieved.
+	Page Page
+	// The page of resources returned (resources = assets)
+	Resources []domain.Asset
+}
+
+// Page represents the JSON object Nexpose provides to help paginate through all the Assets
+type Page struct {
+	// The index (zero-based) of the current page returned.
+	Number int
+	// The maximum size of the page returned.
+	Size int
+	// The total number of pages available.
+	TotalPages int
+	// The total number of resources available across all pages.
+	TotalResources int
+}
+
+// NexposeAssetFetcher is used to create a new
+type NexposeAssetFetcher struct {
+	// The Nexpose host that points to your instance
+	Host string
+	// The number of assets that should be returned at one time
+	PageSize int
+}
+
+// FetchAssets gets all the assets for a given site ID from Nexpose, with pagination.
+// It returns a channel of assets and a channel of errors that can by asynchronously listened to.
+func (c *NexposeAssetFetcher) FetchAssets(ctx context.Context, siteID string) (<-chan domain.Asset, <-chan error) {
+	errChan := make(chan error, 1)
+	assetChan := make(chan domain.Asset, 1)
+	defer close(errChan)
+	defer close(assetChan)
+
+	// have to page through to get all the assets, start with page 0
+	// make the first call to Nexpose to get the total number of pages we'll
+	var currentPage = 0
+	req, err := newNexposeSiteAssetsRequest(c.Host, siteID, currentPage, c.PageSize)
+	if err != nil {
+		errChan <- err
+		return assetChan, errChan
```

Comment:
as opposed to returning `assetChan` it may make more sense to `return nil, errChan` that way we could remove creating the `assetChan` just for purpose of returning in the event we receive an error

Reply:
I agree. I'll make that change